services/smtp-d.go

```diff
@@ -0,0 +1,112 @@
+/*
+* Honeytrap
+* Copyright (C) 2016-2017 DutchSec (https://dutchsec.com/)
+*
+* This program is free software; you can redistribute it and/or modify it under
+* the terms of the GNU Affero General Public License version 3 as published by the
+* Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+* FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
+* details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* version 3 along with this program in the file "LICENSE".  If not, see
+* <http://www.gnu.org/licenses/agpl-3.0.txt>.
+*
+* See https://honeytrap.io/ for more details. All requests should be sent to
+* licensing@honeytrap.io
+*
+* The interactive user interfaces in modified source and object code versions
+* of this program must display Appropriate Legal Notices, as required under
+* Section 5 of the GNU Affero General Public License version 3.
+*
+* In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+* these Appropriate Legal Notices must retain the display of the "Powered by
+* Honeytrap" logo and retain the original copyright notice. If the display of the
+* logo is not reasonably feasible for technical reasons, the Appropriate Legal Notices
+* must display the words "Powered by Honeytrap" and retain the original copyright notice.
+ */
+package services
+
+import (
+	"crypto/tls"
+	"net"
+	"sync"
+)
+
+type Handler interface {
+	Serve(msg Message) error
+}
+
+type HandlerFunc func(msg Message) error
+
+type ServeMux struct {
+	m  []HandlerFunc
+	mu sync.RWMutex
+}
+
+func (mux *ServeMux) HandleFunc(handler func(msg Message) error) {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+
+	mux.m = append(mux.m, handler)
+}
+
+func HandleFunc(handler func(msg Message) error) *ServeMux {
+	DefaultServeMux.HandleFunc(handler)
+	return DefaultServeMux
+}
+
+var DefaultServeMux = NewServeMux()
+
+func NewServeMux() *ServeMux { return &ServeMux{m: make([]HandlerFunc, 0)} }
+
+type Server struct {
+	Banner    string
+	TLSConfig *tls.Config
+	Handler   Handler
+}
+
+func New() (*Server, error) {
```

Comment:
We don't need error handling here, if so, we can have a MustNew, which will panic on error.